modtools/fire-rate.lua

```diff
@@ -0,0 +1,511 @@
+Fort Mode uses a unit's think_counter. After firing a shot, it's set to 80 (not
+accounting for skill stuff). This delays them from acting again until then.
+
+If you make it so onProjItemCheckMovement sets their think_counter to 0 whenever
+they fire something, they can fire 1 shot every tick.
+
+Adventure mode uses think_counter + a fixed delay. After firing a shot, the game
+advances by the fixed delay, then starts decrementing the think_counter. The
+fixed delay is either 10 or 9 from testing. It probably exists to prevent units
+walking into their own shots.
+
+Example:
+- Throw - think_counter 1, total time passed 10
+- Spit - think_counter 1, total time passed 10
+
+When in tactical mode, the fixed delay always occurs after making a ranged
+attack before switching control to the next character.
+
+Then, the think counter delay prevents switching back into that character until
+it's passed.
+
+Some figures might be off by 1 because of delays between firing and when
+onProjItemCheckMovement triggers.
+]]
+
+---------------------------------------------------------------------
+local utils = require("utils")
+local eventful = require("plugins.eventful")
+
+local validArgs = utils.invert({
+  "help",
+  "item",
+  "throw",
+  "material",
+  "fortBase",
+  "advBase",
+  "advSkillFactor",
+  "fortSkillFactor",
+  "fortCap",
+  "advCap",
+  "mode",
+})
+
+-- The (hopefully unique) separator used to split a weapon token and material
+-- token when they're combined to make a key
+local separator = "|||"
+
+-- The (hopefully unique) key used as a substitute for a weapon token when the
+-- attack is coming from a throw.
+local throw_key = "throw"
+
+---------------------------------------------------------------------
+weapon_data = weapon_data or {}
+config = config or {
+  skill_mode = "basic",
+}
+---------------------------------------------------------------------
+
+function register_weapon(weapon_token, material_token, data)
+  local key = weapon_token
+
+  if ( material_token ~= nil ) then
+    key = key ..  separator .. material_token
+  end
+
+  if ( data == nil ) then
+    data = {}
+  end
+
+  local new_entry = {
+    fort_rate = data.fort_rate or 80,
+    adv_rate = data.adv_rate or 80,
+    fort_skill_factor = data.fort_skill_factor or 2.7,
+    adv_skill_factor = data.adv_skill_factor or 2.7,
+  }
+
+  -- Determine caps
+  if ( data.fort_cap ) then
+    new_entry.fort_cap = data.fort_cap
+  else -- default to half main rate
+    -- Half the standard rate, then round to nearest number
+    new_entry.fort_cap = math.floor(new_entry.fort_rate / 2 + 0.5)
+  end
+
+  -- Accounting for the adventure hard-coded cap won't be done at this stage
+  if ( data.adv_cap ) then
+    new_entry.adv_cap = data.adv_cap
+  else
+    new_entry.adv_cap = math.floor(new_entry.adv_rate / 2 + 0.5)
+  end
+
+  weapon_data[key] = new_entry
+  return new_entry
+end
+
+function get_weapon_data(weapon_token, material_token)
```

Comment:
This function could be considerably shorter:
```
return weapon_data[weapon_token..separator..(material_token or '')] or
        weapon_data[weapon_token] or
        register_weapon(weapon_token)
```
with all comments becoming comments about the function as a whole

Reply:
I will usually go out of my way to make things longer than necessary if it means making the code easier to follow, though since most of the clarity comes from the comments in this case, I'm fine making this change (and have done so).